man/leapctl.8.ronn
leapctl(8) -- Administrative utility for privleap
=============================================

<!--
# copyright (c) 2025 - 2025 encrypted support llc <adrelanos@whonix.org>
# see the file copying for copying conditions.
-->

## SYNOPSIS
`leapctl (--create|--destroy) username|uid`

## DESCRIPTION
Provides administrative controls for privleap. Right now this is only used to
open and close sockets used to allow individual user accounts to communicate
with the privleapd server.

You usually will not need to run this directly - if `privleap` was installed
from your distribution's repository, it should be pre-configured to create and
destroy user communication sockets on user login and logout.

## OPTIONS

* `--create` :
  Instructs privleapd to create a communication socket for the specified user.
  This will allow the user to communicate with privleapd and request
  privileged operations to be run.

* `--destroy` :
  Instructs privleapd to destroy a communication socket for the specified
  user. This will prevent the user from communicating with privleapd any
  further.

## ENVIRONMENT VARIABLES
None.

## CONFIG FILES
See privleap-conf.d(5).

## EXIT CODES
0: The requested operation was successful, or did not need to be done.

1: The requested operation failed.

2: privleapd was requested to create a socket for a user, but the specified
user is not permitted to have a comm socket.

## EXAMPLES
Create a communication socket for user `user`:

`leapctl --create user`

Destroy a communication socket for the user with UID 1001:

`leapctl --destroy 1001`

## NOTES
leapctl communicates with privleapd using a UNIX socket that by default can only
be read from and written to by `root`. You will therefore usually have to run
leapctl as root for it to work.

## WWW
https://github.com/Kicksecure/privleap

## AUTHOR
This man page has been written by Aaron Rainbolt (arraybolt3@ubuntu.com).


debian/changelog
privleap (3:3.3-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Wed, 09 Apr 2025 15:15:39 +0000

privleap (3:3.2-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Tue, 08 Apr 2025 14:08:04 +0000

privleap (3:3.1-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Thu, 03 Apr 2025 17:01:21 +0000

privleap (3:3.0-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Mon, 31 Mar 2025 10:20:10 +0000

privleap (3:2.9-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Tue, 18 Mar 2025 10:56:42 +0000

privleap (3:2.8-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Mon, 17 Mar 2025 08:53:20 +0000

privleap (3:2.7-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Sat, 15 Mar 2025 10:52:06 +0000

privleap (3:2.6-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Mon, 10 Mar 2025 07:42:11 +0000

privleap (3:2.5-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Wed, 05 Mar 2025 11:14:15 +0000

privleap (3:2.4-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Mon, 03 Mar 2025 11:24:43 +0000

privleap (3:2.3-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Mon, 03 Mar 2025 10:55:27 +0000

privleap (3:2.2-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Mon, 24 Feb 2025 07:48:03 +0000

privleap (3:2.1-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Fri, 21 Feb 2025 09:55:18 +0000

privleap (3:2.0-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Fri, 21 Feb 2025 09:47:39 +0000

privleap (3:1.9-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Thu, 20 Feb 2025 05:14:34 +0000

privleap (3:1.8-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Mon, 17 Feb 2025 12:28:12 +0000

privleap (3:1.7-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Fri, 14 Feb 2025 00:07:07 +0000

privleap (3:1.6-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Mon, 10 Feb 2025 11:18:56 +0000

privleap (3:1.5-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Sun, 09 Feb 2025 10:03:56 +0000

privleap (3:1.4-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Sat, 08 Feb 2025 11:29:31 +0000

privleap (3:1.3-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Sat, 08 Feb 2025 10:49:43 +0000

privleap (3:1.2-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Fri, 07 Feb 2025 11:17:21 +0000

privleap (3:1.1-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Thu, 06 Feb 2025 17:16:14 +0000

privleap (3:1.0-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Tue, 04 Feb 2025 09:48:35 +0000

privleap (3:0.9-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Sat, 01 Feb 2025 13:38:36 +0000

privleap (3:0.8-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Thu, 30 Jan 2025 16:43:51 +0000

privleap (3:0.7-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Wed, 29 Jan 2025 14:02:47 +0000

privleap (3:0.6-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Wed, 29 Jan 2025 13:13:32 +0000

privleap (3:0.5-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Fri, 24 Jan 2025 13:29:41 +0000

privleap (3:0.4-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Thu, 23 Jan 2025 16:26:21 +0000

privleap (3:0.3-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Thu, 23 Jan 2025 14:51:55 +0000

privleap (3:0.2-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Patrick Schleizer <adrelanos@whonix.org>  Wed, 22 Jan 2025 14:39:25 +0000

privleap (3:0.1-1) unstable; urgency=medium

  * New upstream version (local package).

 -- Aaron Rainbolt <arraybolt3@ubuntu.com>  Sat, 18 Jan 2025 22:04:13 -0600


usr/lib/python3/dist-packages/privleap/leapctl.py
#!/usr/bin/python3 -u

# Copyright (C) 2025 - 2025 ENCRYPTED SUPPORT LLC <adrelanos@whonix.org>
# See the file COPYING for copying conditions.

# pylint: disable=broad-exception-caught,duplicate-code
# Rationale:
#   broad-exception-caught: except blocks are general error handlers.
#   duplicate-code: This is being triggered because of generic_error and
#     unexpected_error_msg, which are very similar in leapctl and leaprun but
#     can't be reasonably broken out of either due to the fact that they access
#     incompatible variants of the cleanup_and_exit function.

"""leapctl.py - privleapd client for controlling available comm sockets."""

import sys
from typing import NoReturn

# import privleap as pl

import privleap.privleap as pl


# pylint: disable=too-few-public-methods
# Rationale:
#   too-few-public-methods: This class just stores global variables, it needs no
#     public methods. Namespacing global variables in a class makes things
#     safer.
class LeapctlGlobal:
    """
    Global variables for leapctl.
    """

    control_session: pl.PrivleapSession | None = None


def cleanup_and_exit(exit_code: int) -> NoReturn:
    """
    Ensures any active session is closed, then exits.
    """

    if LeapctlGlobal.control_session is not None:
        LeapctlGlobal.control_session.close_session()
    sys.exit(exit_code)


def print_usage() -> NoReturn:
    """
    Prints usage information.
    """

    print(
        """leapctl <--create|--destroy> <user>
leapctl --reload

    --create : Specifies that leapctl should request a communication socket to
               be created for the specified user account.
    --destroy : Specifies that leapctl should request a communication socket
                to be destroyed for the specified user account.
    --reload : Instructs privleapd to reload configuration without restarting.
    user : The username or UID of the user account to create or destroy a
           communication socket for."""
    )
    cleanup_and_exit(1)


def generic_error(error_msg: str, exit_int: int = 1) -> NoReturn:
    """
    Prints an error, then cleans up and exits.
    """

    print(f"ERROR: {error_msg}", file=sys.stderr)
    cleanup_and_exit(exit_int)


def unexpected_msg_error(control_msg: pl.PrivleapMsg) -> NoReturn:
    """
    Alerts about an unexpected message type being received from the server.
    """

    print(
        "ERROR: privleapd returned unexpected message as follows:",
        file=sys.stderr,
    )
    print(control_msg.serialize(), file=sys.stderr)
    cleanup_and_exit(1)


def start_control_session() -> None:
    """
    Starts a control session with the server.
    """

    try:
        LeapctlGlobal.control_session = pl.PrivleapSession(
            is_control_session=True
        )
    except Exception:
        generic_error("Could not connect to privleapd!")


def handle_create_request(user_name: str) -> NoReturn:
    """
    Requests that privleapd create a socket for the specified user, and handles
      the response to the request.
    """

    assert LeapctlGlobal.control_session is not None

    try:
        LeapctlGlobal.control_session.send_msg(
            pl.PrivleapControlClientCreateMsg(user_name)
        )
    except Exception:
        generic_error("Could not request privleapd to create a comm socket!")

    try:
        control_msg: pl.PrivleapMsg = LeapctlGlobal.control_session.get_msg()
    except Exception:
        generic_error("privleapd didn't return a valid response!")

    # noinspection PyUnboundLocalVariable
    if isinstance(control_msg, pl.PrivleapControlServerOkMsg):
        print(f"Comm socket created for account {repr(user_name)}.")
        cleanup_and_exit(0)
    elif isinstance(control_msg, pl.PrivleapControlServerControlErrorMsg):
        generic_error(
            "privleapd encountered an error while creating a comm socket "
            f"for account {repr(user_name)}!"
        )
    elif isinstance(control_msg, pl.PrivleapControlServerDisallowedUserMsg):
        generic_error(
            f"Account {repr(user_name)} is not permitted to have a comm socket!",
            2,
        )
    elif isinstance(
        control_msg,
        pl.PrivleapControlServerExpectedDisallowedUserMsg,
    ):
        print(
            f"Account {repr(user_name)} is not permitted to have a comm "
            "socket, as expected, ok."
        )
        cleanup_and_exit(0)
    elif isinstance(control_msg, pl.PrivleapControlServerExistsMsg):
        print(f"Comm socket already exists for account {repr(user_name)}.")
        cleanup_and_exit(0)
    else:
        unexpected_msg_error(control_msg)


def handle_destroy_request(user_name: str) -> NoReturn:
    """
    Requests that privleapd destroy a socket for the specified user, and handles
      the response to the request.
    """

    assert LeapctlGlobal.control_session is not None

    try:
        # noinspection PyUnboundLocalVariable
        LeapctlGlobal.control_session.send_msg(
            pl.PrivleapControlClientDestroyMsg(user_name)
        )
    except Exception:
        generic_error("Could not request privleapd to destroy a comm socket!")

    try:
        control_msg: pl.PrivleapMsg = LeapctlGlobal.control_session.get_msg()
    except Exception:
        generic_error("privleapd didn't return a valid response!")

    # noinspection PyUnboundLocalVariable
    if isinstance(control_msg, pl.PrivleapControlServerOkMsg):
        print(f"Comm socket destroyed for account {repr(user_name)}.")
        cleanup_and_exit(0)
    elif isinstance(control_msg, pl.PrivleapControlServerControlErrorMsg):
        generic_error(
            "privleapd encountered an error while destroying a comm "
            f"socket for account {repr(user_name)}!"
        )
    elif isinstance(control_msg, pl.PrivleapControlServerNouserMsg):
        print(f"Comm socket does not exist for account {repr(user_name)}.")
        cleanup_and_exit(0)
    elif isinstance(control_msg, pl.PrivleapControlServerPersistentUserMsg):
        print(
            f"Cannot destroy socket for persistent account {repr(user_name)}."
        )
        # It is not an error to try to destroy a socket for a persistent user,
        # since this may legitimately happen if someone logs in as a user that
        # happens to be persistent in privleap's config, and then logs out.
        cleanup_and_exit(0)
    else:
        unexpected_msg_error(control_msg)


def handle_reload_request() -> NoReturn:
    """
    Requests that privleapd reload its configuration files without restarting.
    """

    assert LeapctlGlobal.control_session is not None

    try:
        LeapctlGlobal.control_session.send_msg(
            pl.PrivleapControlClientReloadMsg()
        )
    except Exception:
        generic_error("Could not request privleapd to reload configuration!")

    try:
        control_msg: pl.PrivleapMsg = LeapctlGlobal.control_session.get_msg()
    except Exception:
        generic_error("privleapd didn't return a valid response!")

    if isinstance(control_msg, pl.PrivleapControlServerOkMsg):
        print("privleapd configuration reload successful.")
        cleanup_and_exit(0)
    elif isinstance(control_msg, pl.PrivleapControlServerControlErrorMsg):
        generic_error("privleapd failed to reload configuration!")
    else:
        unexpected_msg_error(control_msg)


def main() -> NoReturn:
    """
    Main function.
    """

    if len(sys.argv) < 2 or len(sys.argv) > 3:
        print_usage()

    control_action: str = sys.argv[1]
    control_user: str | None = None
    if len(sys.argv) == 3:
        control_user = sys.argv[2]

    if control_action not in ("--create", "--destroy", "--reload"):
        print_usage()
    if control_action == "--reload" and control_user is not None:
        print_usage()

    if control_user is not None:
        orig_control_user = control_user
        control_user = pl.PrivleapCommon.normalize_user_id(control_user)
        # Allow a username that doesn't exist to be passed when using --destroy,
        # so if a user is deleted before their comm socket is destroyed, the
        # socket can be destroyed anyway.
        if control_user is None:
            if control_action != "--destroy":
                generic_error("Specified user account does not exist.")
            else:
                control_user = orig_control_user

    start_control_session()

    if control_action == "--create":
        assert control_user is not None
        handle_create_request(control_user)
    elif control_action == "--destroy":
        assert control_user is not None
        handle_destroy_request(control_user)
    else:
        handle_reload_request()


if __name__ == "__main__":
    main()
